app/code/core/Mage/Usa/Model/Shipping/Carrier/Ups.php

```diff
@@ -1584,13 +1455,403 @@ protected function _sendShipmentAcceptRequest(SimpleXMLElement $shipmentConfirmR
+            curl_setopt($ch, CURLOPT_TIMEOUT, 30);
+            curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, $this->getConfigFlag('verify_peer'));
+            $xmlResponse = curl_exec($ch);
+
+            $debugData['result'] = $xmlResponse;
+            $this->_setCachedQuotes($xmlRequest, $xmlResponse);
+        } catch (Exception $e) {
+            $debugData['result'] = ['error' => $e->getMessage(), 'code' => $e->getCode()];
+            $xmlResponse = '';
+        }
+
+        try {
+            $response = new SimpleXMLElement($xmlResponse);
+        } catch (Exception $e) {
+            $debugData['result'] = ['error' => $e->getMessage(), 'code' => $e->getCode()];
+        }
+
+        $result = new Varien_Object();
+        if (isset($response->Error)) {
+            $result->setErrors((string)$response->Error->ErrorDescription);
+        } else {
+            $shippingLabelContent = (string)$response->ShipmentResults->PackageResults->LabelImage->GraphicImage;
+            $trackingNumber       = (string)$response->ShipmentResults->PackageResults->TrackingNumber;
+
+            $result->setShippingLabelContent(base64_decode($shippingLabelContent));
+            $result->setTrackingNumber($trackingNumber);
+        }
+
+        $this->_debug($debugData);
+        return $result;
+    }
+
+    protected function _doShipmentRequest(Varien_Object $request): Varien_Object
+    {
+        // this "if" will be removed after XML APIs will be shut down
+        if ($this->getConfigData('type') == 'UPS_XML') {
+            return $this->_doShipmentRequestXML($request);
+        }
+
+        // REST is default
+        return $this->_doShipmentRequestRest($request);
+    }
+
     /**
      * Do shipment request to carrier web service, obtain Print Shipping Labels and process errors in response
+     */
+    protected function _doShipmentRequestRest(Varien_Object $request): Varien_Object
+    {
+        $request->setShipperAddressCountryCode(
+            $this->getNormalizedCountryCode(
+                $request->getShipperAddressCountryCode(),
+                $request->getShipperAddressStateOrProvinceCode(),
+                $request->getShipperAddressPostalCode(),
+            )
+        );
+
+        $request->setRecipientAddressCountryCode(
+            $this->getNormalizedCountryCode(
+                $request->getRecipientAddressCountryCode(),
+                $request->getRecipientAddressStateOrProvinceCode(),
+                $request->getRecipientAddressPostalCode(),
+            )
+        );
+
+        $result = new Varien_Object();
+        $this->_prepareShipmentRequest($request);
+        $rawJsonRequest = $this->_formShipmentRestRequest($request);
+        $accessToken = $this->setAPIAccessRequest();
+        $this->_debug(['request_quote' => $rawJsonRequest]);
+
+        if ($this->getConfigFlag('mode_xml')) {
+            $shipConfirmUrl = $this->_liveUrls['ShipRestConfirm'];
```

Comment:
Why do some URLs use these hardcoded "default" vs "live" URLs, but some use the config.xml approach? Specifically, in for shipconfirm, the XML code checks `$this->getConfig('shipconfirm_xml_url')` and then falls back to `$this->_defaultUrls['ShipConfirm']`. Why doesn't this code add a config for 'shipconfirm_rest_url', and why doesn't the XML code ever use the `_liveUrls['ShipConfirm']` value?

Reply:
I've no idea why it was done this way but it's how the original module was done and how the M2 patch treats it

Reply:
Fair enough. Upon looking at the Magento2 code, I see the similar approach.

However, I still think what's done here doesn't match M2. The M2 code has methods `getShipConfirmUrl()`, and `getShipAcceptUrl()`. `getShipConfirmUrl()` does seem to grab a config value for the url before falling back to the live vs default URLs:

```
    public function getShipConfirmUrl()
    {
        $url = $this->getConfigData('url');
        if (!$url) {
            if ($this->getConfigData('is_account_live')) {
                if ($this->getConfigData('type') == 'UPS_XML') {
                    $url = $this->_liveUrls['ShipConfirm'];
                } else {
                    $url = $this->_liveUrls['ShipRestConfirm'];
                }

                return $url;
            } else {
                if ($this->getConfigData('type') == 'UPS_XML') {
                    $url = $this->_defaultUrls['ShipConfirm'];
                } else {
                    $url = $this->_defaultUrls['ShipRestConfirm'];
                }

                return $url;
            }
        }

        return $url;
    }
```

Reply:
I lost a bit of control here, but this patch is already working, you could make suggestions to the code or a PR to my repo if you want because this way it's harder for me to check.